modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/CheckCatalogVersionOnActionRequest.java

```diff
@@ -59,20 +72,53 @@ public CheckCatalogVersionOnActionRequest(CatalogService catalogService) {
             if (!isMetadataAvailableFor(requiredCatalogVersion, catalogService)) {
                 // TODO: IGNITE-20298 - throttle logging.
                 LOG.warn(
-                        "Metadata not yet available, group {}, required level {}; rejecting ActionRequest with EBUSY.",
+                        "Metadata not yet available, rejecting ActionRequest with EBUSY [group={}, requiredLevel={}].",
                         request.groupId(), requiredCatalogVersion
                 );
 
                 return RaftRpcFactory.DEFAULT //
                     .newResponse(
                             node.getRaftOptions().getRaftMessagesFactory(),
                             RaftError.EBUSY,
-                            "Metadata not yet available, group '%s', required level %d; rejecting ActionRequest with EBUSY.",
+                            "Metadata not yet available, rejecting ActionRequest with EBUSY [group=%s, requiredLevel=%d].",
                             request.groupId(), requiredCatalogVersion
                     );
             }
         }
 
         return null;
     }
+
+    /**
+     * Builds an {@link org.apache.ignite.raft.jraft.rpc.RpcRequests.ErrorResponse} if the current node is not a leader.
+     * It tries to mirror the behavior of {@link NodeImpl} in this regard.
+     *
+     * <p>NB that it's important to return leaderId for the 'not a leader and not transferring leadership' state
+     * so that the caller switches to the actual leader.
+     *
+     * @param node The node.
+     * @return ErrorResponse if this node is not a leader, {@code null} otherwise.
+     */
+    private static @Nullable Message errorResponseIfNotLeader(Node node) {
+        State state = node.getNodeState();
+
+        if (state == State.STATE_LEADER) {
+            return null;
+        }
+
+        Status st = NodeImpl.cannotApplyBecauseNotLeaderStatus(state);
+
+        PeerId leaderId = node.getLeaderId();
+
+        if (st.getRaftError() == RaftError.EPERM && leaderId != null) {
```

Comment:
What does this mean? Please leave a comment that it's "not a leader" status.
By the way, why is it necessary to check the code? Is it not enough to have a leaderId check?
Last question, why do we create responses differently in these cases?

Reply:
I added a comment and simplified the code.

As for 'why don't just return leaderId if we have it', take a look at `IgniteCliRpcRequestClosure:74`: they only enrich the response with leaderId for EPERM. I did not want to change this to avoid breaking this protocol.

Reply:
That's crazy, thank you for the explanation